src/irmin/type/type_binary.ml

```diff
@@ -232,162 +253,208 @@ module Decode = struct
     | `Int -> int buf ofs
     | `Int8 -> int8 buf ofs
     | `Int16 -> int16 buf ofs
-    | `Int32 -> int32 buf ofs >|= Int32.to_int
-    | `Int64 -> int64 buf ofs >|= Int64.to_int
-    | `Fixed n -> ok ofs n
+    | `Int32 ->
+        let ofs, i = int32 buf ofs in
+        (ofs, Int32.to_int i)
+    | `Int64 ->
+        let ofs, i = int64 buf ofs in
+        (ofs, Int64.to_int i)
+    | `Fixed n -> (ofs, n)
+    | `Unboxed -> (ofs, String.length buf - ofs)
+
+  let mk_unboxed of_string of_bytes _ buf ofs =
+    let len = String.length buf - ofs in
+    if ofs = 0 then (len, of_string buf)
+    else
+      let str = Bytes.create len in
+      String.blit buf ofs str 0 len;
+      (ofs + len, of_bytes str)
 
-  let string ?(headers = true) n buf ofs =
-    let sub ofs len =
-      if ofs = 0 && len = String.length buf then ok len buf
-      else
-        let str = Bytes.create len in
-        String.blit buf ofs str 0 len;
-        ok (ofs + len) (Bytes.unsafe_to_string str)
-    in
-    match (headers, n) with
-    | _, `Fixed len -> sub ofs len
-    | true, _ -> len buf ofs n >>= fun (ofs, len) -> sub ofs len
-    | false, _ -> sub ofs (String.length buf - ofs)
-
-  let bytes ?(headers = true) n buf ofs =
-    let sub ofs len =
-      if ofs = 0 && len = String.length buf then ok len (Bytes.of_string buf)
+  let mk_boxed of_string of_bytes n =
+    let sub len buf ofs =
+      if ofs = 0 && len = String.length buf then (len, of_string buf)
       else
         let str = Bytes.create len in
         String.blit buf ofs str 0 len;
```

Comment:
Not related to this PR, but it might be worth failing more gracefully here in the case `len > String.length buf` (which corresponds to trying to read a header that doesn't exist / malformed length header). That's quite easy to enounter as a user of `Type.Unboxed`.